File: analyze_sync_compatibility.py
#!/usr/bin/env python3
"""
تحليل مطابقة Sync Rules مع PowerSync Schema و Supabase
============================================================
هذا السكريبت يقارن:
1. Sync Rules (powersync-sync-rules.yaml)
2. PowerSync Schema (PowerSyncSchema.ts)
3. Supabase Schema (من supabase.ts)
"""

import re
import json
from pathlib import Path
from typing import Dict, List, Set, Tuple

# مسارات الملفات
BASE_DIR = Path(__file__).parent
SYNC_RULES_PATH = BASE_DIR / "powersync-sync-rules.yaml"
SCHEMA_PATH = BASE_DIR / "src/lib/powersync/PowerSyncSchema.ts"
SUPABASE_PATH = BASE_DIR / "src/types/supabase.ts"

def extract_sync_rules_tables(content: str) -> Dict[str, List[str]]:
    """استخراج الجداول والأعمدة من Sync Rules"""
    tables = {}
    lines = content.split('\n')
    
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        
        # البحث عن SELECT statements
        if line.startswith('- SELECT'):
            columns = []
            i += 1
            
            # جمع الأعمدة حتى نصل إلى FROM
            while i < len(lines):
                line = lines[i].strip()
                
                if line.startswith('FROM'):
                    # استخراج اسم الجدول
                    match = re.search(r'FROM\s+(\w+)', line)
                    if match:
                        table_name = match.group(1)
                        # تنظيف الأعمدة
                        clean_columns = []
                        for col in columns:
                            # تقسيم الأعمدة المفصولة بفواصل
                            for c in col.split(','):
                                c = c.strip()
                                if c and not c.startswith('#') and not c.startswith('--'):
                                    clean_columns.append(c)
                        tables[table_name] = clean_columns
                    break
                
                if line and not line.startswith('#') and not line.startswith('WHERE'):
                    columns.append(line)
                
                i += 1
        i += 1
    
    return tables

def extract_powersync_schema(content: str) -> Dict[str, List[str]]:
    """استخراج الجداول من PowerSync Schema"""
    tables = {}
    
    # البحث عن جميع تعريفات الجداول
    pattern = r'const\s+(\w+)\s*=\s*new\s+Table\s*\(\s*\{([^}]+)\}'
    
    matches = re.finditer(pattern, content, re.DOTALL)
    
    for match in matches:
        table_name = match.group(1)
        table_content = match.group(2)
        
        # استخراج الأعمدة
        columns = []
        column_pattern = r'(\w+):\s*column\.\w+'
        
        for col_match in re.finditer(column_pattern, table_content):
            columns.append(col_match.group(1))
        
        # تحويل اسم الجدول إلى snake_case
        snake_name = re.sub(r'(?<!^)(?=[A-Z])', '_', table_name).lower()
        tables[snake_name] = columns
        tables[table_name] = columns  # حفظ الاسم الأصلي أيضاً
    
    return tables

def extract_supabase_schema(content: str) -> Dict[str, List[str]]:
    """استخراج الجداول من Supabase Types"""
    tables = {}
    
    # البحث عن تعريفات الجداول
    pattern = r'(\w+):\s*\{[^}]*Row:\s*\{([^}]+)\}'
    
    matches = re.finditer(pattern, content, re.DOTALL)
    
    for match in matches:
        table_name = match.group(1)
        row_content = match.group(2)
        
        # استخراج الأعمدة
        columns = []
        column_pattern = r'(\w+):\s*(?:string|number|boolean|Json|\w+\s*\|\s*null|\w+\s*\[\]|\w+\s*\[\]\s*\|\s*null)'
        
        for col_match in re.finditer(column_pattern, row_content):
            columns.append(col_match.group(1))
        
        tables[table_name] = columns
    
    return tables

def normalize_column_name(col: str) -> str:
    """تطبيع اسم العمود"""
    return col.strip().lower()

def compare_schemas():
    """المقارنة الرئيسية"""
    print("🔍 بدء التحليل...\n")
    
    # قراءة الملفات
    sync_rules_content = SYNC_RULES_PATH.read_text(encoding='utf-8')
    schema_content = SCHEMA_PATH.read_text(encoding='utf-8')
    supabase_content = SUPABASE_PATH.read_text(encoding='utf-8')
    
    # استخراج البيانات
    sync_rules_tables = extract_sync_rules_tables(sync_rules_content)
    power_sync_tables = extract_powersync_schema(schema_content)
    supabase_tables = extract_supabase_schema(supabase_content)
    
    print(f"📊 إحصائيات:")
    print(f"   - Sync Rules: {len(sync_rules_tables)} جدول")
    print(f"   - PowerSync Schema: {len(power_sync_tables)} جدول")
    print(f"   - Supabase Schema: {len(supabase_tables)} جدول\n")
    
    issues = []
    warnings = []
    matches = []
    
    # مقارنة كل جدول في Sync Rules
    for table_name, sync_columns in sync_rules_tables.items():
        print(f"\n📋 جدول: {table_name}")
        print(f"   📊 عدد الأعمدة في Sync Rules: {len(sync_columns)}")
        
        # البحث عن الجدول في PowerSync Schema
        power_sync_name = None
        power_sync_columns = None
        
        # البحث بالاسم المباشر أو snake_case
        if table_name in power_sync_tables:
            power_sync_name = table_name
            power_sync_columns = power_sync_tables[table_name]
        else:
            # البحث عن تطابق في الأسماء
            for ps_name in power_sync_tables.keys():
                if ps_name.lower() == table_name.lower() or ps_name.replace('_', '') == table_name.replace('_', ''):
                    power_sync_name = ps_name
                    power_sync_columns = power_sync_tables[ps_name]
                    break
        
        # البحث عن الجدول في Supabase
        supabase_columns = supabase_tables.get(table_name)
        
        if not power_sync_name:
            issue = f"❌ الجدول '{table_name}' غير موجود في PowerSync Schema"
            issues.append(issue)
            print(f"   ⚠️  {issue}")
        else:
            print(f"   ✅ موجود في PowerSync Schema باسم: {power_sync_name}")
            
            # مقارنة الأعمدة
            sync_cols_normalized = {normalize_column_name(c): c for c in sync_columns}
            ps_cols_normalized = {normalize_column_name(c): c for c in power_sync_columns}
            
            missing_in_schema = []
            for sync_col_norm, sync_col_orig in sync_cols_normalized.items():
                if sync_col_norm not in ps_cols_normalized:
                    missing_in_schema.append(sync_col_orig)
            
            extra_in_schema = []
            for ps_col_norm, ps_col_orig in ps_cols_normalized.items():
                if ps_col_norm not in sync_cols_normalized:
                    extra_in_schema.append(ps_col_orig)
            
            if missing_in_schema:
                issue = f"❌ أعمدة مفقودة في PowerSync Schema للجدول '{table_name}': {', '.join(missing_in_schema)}"
                issues.append(issue)
                print(f"   ⚠️  {issue}")
            
            if extra_in_schema:
                warning = f"⚠️  أعمدة إضافية في PowerSync Schema للجدول '{table_name}': {', '.join(extra_in_schema)}"
                warnings.append(warning)
                print(f"   ℹ️  {warning}")
            
            print(f"   📊 عدد الأعمدة في Schema: {len(power_sync_columns)}")
        
        if not supabase_columns:
            issue = f"❌ الجدول '{table_name}' غير موجود في Supabase"
            issues.append(issue)
            print(f"   ⚠️  {issue}")
        else:
            print(f"   ✅ موجود في Supabase")
            
            # مقارنة الأعمدة مع Supabase
            sync_cols_normalized = {normalize_column_name(c): c for c in sync_columns}
            supabase_cols_normalized = {normalize_column_name(c): c for c in supabase_columns}
            
            missing_in_supabase = []
            for sync_col_norm, sync_col_orig in sync_cols_normalized.items():
                if sync_col_norm not in supabase_cols_normalized:
                    missing_in_supabase.append(sync_col_orig)
            
            if missing_in_supabase:
                issue = f"❌ أعمدة مفقودة في Supabase للجدول '{table_name}': {', '.join(missing_in_supabase)}"
                issues.append(issue)
                print(f"   ⚠️  {issue}")
            
            print(f"   📊 عدد الأعمدة في Supabase: {len(supabase_columns)}")
            
            # التحقق من المطابقة الكاملة
            if power_sync_columns and supabase_columns:
                sync_set = {normalize_column_name(c) for c in sync_columns}
                ps_set = {normalize_column_name(c) for c in power_sync_columns}
                supabase_set = {normalize_column_name(c) for c in supabase_columns}
                
                if sync_set == ps_set == supabase_set:
                    matches.append(table_name)
                    print(f"   ✅ مطابق تماماً!")
    
    # تقرير نهائي
    print('\n\n' + '=' * 80)
    print('📊 تقرير التحليل النهائي')
    print('=' * 80)
    
    if issues:
        print(f"\n❌ المشاكل ({len(issues)}):")
        for issue in issues:
            print(f"   {issue}")
    
    if warnings:
        print(f"\n⚠️  التحذيرات ({len(warnings)}):")
        for warning in warnings:
            print(f"   {warning}")
    
    if matches:
        print(f"\n✅ الجداول المطابقة تماماً ({len(matches)}):")
        for match in matches:
            print(f"   - {match}")
    
    if not issues and not warnings:
        print('\n✅ كل شيء مطابق! لا توجد مشاكل.')
    
    # حفظ التقرير
    report_lines = [
        'تقرير مطابقة Sync Rules مع PowerSync Schema و Supabase',
        '=' * 80,
        f'تاريخ التحليل: {Path(__file__).stat().st_mtime}',
        '',
        f'عدد الجداول في Sync Rules: {len(sync_rules_tables)}',
        f'عدد الجداول في PowerSync Schema: {len(power_sync_tables)}',
        f'عدد الجداول في Supabase: {len(supabase_tables)}',
        '',
    ]
    
    if issues:
        report_lines.append(f'\n❌ المشاكل ({len(issues)}):')
        report_lines.extend([f'   {issue}' for issue in issues])
    
    if warnings:
        report_lines.append(f'\n⚠️  التحذيرات ({len(warnings)}):')
        report_lines.extend([f'   {warning}' for warning in warnings])
    
    if matches:
        report_lines.append(f'\n✅ الجداول المطابقة ({len(matches)}):')
        report_lines.extend([f'   - {match}' for match in matches])
    
    report_path = BASE_DIR / 'sync_compatibility_report.txt'
    report_path.write_text('\n'.join(report_lines), encoding='utf-8')
    print(f'\n💾 تم حفظ التقرير في: {report_path}')
    
    return {
        'issues': issues,
        'warnings': warnings,
        'matches': matches,
        'sync_rules_tables': sync_rules_tables,
        'power_sync_tables': power_sync_tables,
        'supabase_tables': supabase_tables
    }

if __name__ == '__main__':
    try:
        result = compare_schemas()
    except Exception as e:
        print(f'❌ خطأ في التحليل: {e}')
        import traceback
        traceback.print_exc()
        exit(1)

File: analyze_sync_compatibility.js
/**
 * تحليل مطابقة Sync Rules مع PowerSync Schema و Supabase
 * ============================================================
 * هذا السكريبت يقارن:
 * 1. Sync Rules (powersync-sync-rules.yaml)
 * 2. PowerSync Schema (PowerSyncSchema.ts)
 * 3. Supabase Schema (من supabase.ts)
 */

const fs = require('fs');
const path = require('path');

// قراءة ملف Sync Rules
const syncRulesPath = path.join(__dirname, 'powersync-sync-rules.yaml');
const syncRulesContent = fs.readFileSync(syncRulesPath, 'utf8');

// قراءة ملف PowerSync Schema
const schemaPath = path.join(__dirname, 'src/lib/powersync/PowerSyncSchema.ts');
const schemaContent = fs.readFileSync(schemaPath, 'utf8');

// قراءة ملف Supabase Types
const supabasePath = path.join(__dirname, 'src/types/supabase.ts');
const supabaseContent = fs.readFileSync(supabasePath, 'utf8');

// استخراج الجداول والأعمدة من Sync Rules
function extractSyncRulesTables(content) {
  const tables = {};
  const lines = content.split('\n');
  let currentTable = null;
  let inSelect = false;
  let columns = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    
    // البحث عن SELECT statements
    if (line.startsWith('- SELECT')) {
      inSelect = true;
      columns = [];
      continue;
    }
    
    if (inSelect && line.startsWith('FROM')) {
      // استخراج اسم الجدول
      const match = line.match(/FROM\s+(\w+)/);
      if (match) {
        currentTable = match[1];
        tables[currentTable] = columns.filter(col => col && !col.includes('--'));
      }
      inSelect = false;
      continue;
    }
    
    if (inSelect && line && !line.startsWith('#') && !line.startsWith('WHERE')) {
      // استخراج الأعمدة
      const cols = line.split(',').map(col => col.trim()).filter(col => col);
      columns.push(...cols);
    }
  }
  
  return tables;
}

// استخراج الجداول من PowerSync Schema
function extractPowerSyncSchema(content) {
  const tables = {};
  const tableRegex = /const\s+(\w+)\s*=\s*new\s+Table\s*\(/g;
  let match;
  
  while ((match = tableRegex.exec(content)) !== null) {
    const tableName = match[1];
    const startPos = match.index;
    
    // البحث عن نهاية تعريف الجدول
    let braceCount = 0;
    let inTable = false;
    let tableDef = '';
    
    for (let i = startPos; i < content.length; i++) {
      const char = content[i];
      if (char === '(') {
        braceCount++;
        inTable = true;
      }
      if (inTable) tableDef += char;
      if (char === ')') {
        braceCount--;
        if (braceCount === 0) break;
      }
    }
    
    // استخراج الأعمدة
    const columns = [];
    const columnRegex = /(\w+):\s*column\.\w+/g;
    let colMatch;
    while ((colMatch = columnRegex.exec(tableDef)) !== null) {
      columns.push(colMatch[1]);
    }
    
    tables[tableName] = columns;
  }
  
  return tables;
}

// استخراج الجداول من Supabase Types
function extractSupabaseSchema(content) {
  const tables = {};
  const tableRegex = /(\w+):\s*\{[\s\S]*?Row:\s*\{([\s\S]*?)\}/g;
  let match;
  
  while ((match = tableRegex.exec(content)) !== null) {
    const tableName = match[1];
    const rowContent = match[2];
    
    // استخراج الأعمدة من Row type
    const columns = [];
    const columnRegex /(\w+):\s*(?:string|number|boolean|Json|\w+\s*\|\s*null)/g;
    let colMatch;
    while ((colMatch = columnRegex.exec(rowContent)) !== null) {
      columns.push(colMatch[1]);
    }
    
    tables[tableName] = columns;
  }
  
  return tables;
}

// تحويل أسماء الجداول إلى صيغة موحدة
function normalizeTableName(name) {
  // تحويل من camelCase إلى snake_case
  return name.replace(/([A-Z])/g, '_$1').toLowerCase().replace(/^_/, '');
}

// المقارنة الرئيسية
function compareSchemas() {
  console.log('🔍 بدء التحليل...\n');
  
  const syncRulesTables = extractSyncRulesTables(syncRulesContent);
  const powerSyncTables = extractPowerSyncSchema(schemaContent);
  const supabaseTables = extractSupabaseSchema(supabaseContent);
  
  console.log(`📊 إحصائيات:`);
  console.log(`   - Sync Rules: ${Object.keys(syncRulesTables).length} جدول`);
  console.log(`   - PowerSync Schema: ${Object.keys(powerSyncTables).length} جدول`);
  console.log(`   - Supabase Schema: ${Object.keys(supabaseTables).length} جدول\n`);
  
  const issues = [];
  const warnings = [];
  
  // مقارنة كل جدول في Sync Rules
  for (const [tableName, syncColumns] of Object.entries(syncRulesTables)) {
    console.log(`\n📋 جدول: ${tableName}`);
    
    // البحث عن الجدول في PowerSync Schema
    const powerSyncName = Object.keys(powerSyncTables).find(
      name => normalizeTableName(name) === tableName
    );
    
    // البحث عن الجدول في Supabase
    const supabaseName = Object.keys(supabaseTables).find(
      name => name === tableName
    );
    
    if (!powerSyncName) {
      issues.push(`❌ الجدول ${tableName} غير موجود في PowerSync Schema`);
      console.log(`   ⚠️  غير موجود في PowerSync Schema`);
    } else {
      const powerSyncColumns = powerSyncTables[powerSyncName];
      const missingInSchema = syncColumns.filter(col => !powerSyncColumns.includes(col));
      const extraInSchema = powerSyncColumns.filter(col => !syncColumns.includes(col));
      
      if (missingInSchema.length > 0) {
        issues.push(`❌ أعمدة مفقودة في PowerSync Schema للجدول ${tableName}: ${missingInSchema.join(', ')}`);
        console.log(`   ⚠️  أعمدة مفقودة في Schema: ${missingInSchema.join(', ')}`);
      }
      
      if (extraInSchema.length > 0) {
        warnings.push(`⚠️  أعمدة إضافية في PowerSync Schema للجدول ${tableName}: ${extraInSchema.join(', ')}`);
      }
    }
    
    if (!supabaseName) {
      issues.push(`❌ الجدول ${tableName} غير موجود في Supabase`);
      console.log(`   ⚠️  غير موجود في Supabase`);
    } else {
      const supabaseColumns = supabaseTables[supabaseName];
      const missingInSupabase = syncColumns.filter(col => !supabaseColumns.includes(col));
      
      if (missingInSupabase.length > 0) {
        issues.push(`❌ أعمدة مفقودة في Supabase للجدول ${tableName}: ${missingInSupabase.join(', ')}`);
        console.log(`   ⚠️  أعمدة مفقودة في Supabase: ${missingInSupabase.join(', ')}`);
      }
    }
    
    if (powerSyncName && supabaseName) {
      console.log(`   ✅ موجود في كلا الملفين`);
      console.log(`   📊 عدد الأعمدة: Sync Rules (${syncColumns.length}), Schema (${powerSyncTables[powerSyncName].length}), Supabase (${supabaseTables[supabaseName].length})`);
    }
  }
  
  // تقرير نهائي
  console.log('\n\n' + '='.repeat(80));
  console.log('📊 تقرير التحليل النهائي');
  console.log('='.repeat(80));
  
  if (issues.length === 0 && warnings.length === 0) {
    console.log('✅ كل شيء مطابق! لا توجد مشاكل.');
  } else {
    if (issues.length > 0) {
      console.log(`\n❌ المشاكل (${issues.length}):`);
      issues.forEach(issue => console.log(`   ${issue}`));
    }
    
    if (warnings.length > 0) {
      console.log(`\n⚠️  التحذيرات (${warnings.length}):`);
      warnings.forEach(warning => console.log(`   ${warning}`));
    }
  }
  
  return { issues, warnings };
}

// تشغيل التحليل
try {
  const result = compareSchemas();
  
  // حفظ التقرير في ملف
  const reportPath = path.join(__dirname, 'sync_compatibility_report.txt');
  const report = [
    'تقرير مطابقة Sync Rules مع PowerSync Schema و Supabase',
    '='.repeat(80),
    `تاريخ التحليل: ${new Date().toLocaleString('ar-SA')}`,
    '',
    ...result.issues.map(i => i),
    ...result.warnings.map(w => w),
  ].join('\n');
  
  fs.writeFileSync(reportPath, report, 'utf8');
  console.log(`\n💾 تم حفظ التقرير في: ${reportPath}`);
} catch (error) {
  console.error('❌ خطأ في التحليل:', error);
  process.exit(1);
}

























